Adding type aliases (#704)
informalsystems/apalache: docs/src/adr/002adr-types.md

uninterpreted sort in SMT. Essentially, values of an uninterpreted type can
be only checked for equality.

Like in programming languages, a type may contain a reference to a type alias.
This is purely a convenience feature to make type annotations more concise and

> easier to maintain. We expect only the user to write type aliases, whereas
> tools should always exchange with types in the alias-free form.

Reviewer: ```suggestion
easier to maintain. We expect that only users will write type aliases:
tools should always exchange data with types in the alias-free form.
```

**Examples.**

* `x` is an integer. Its type is `Int`.
* `f` is a function from an integer to an integer. Its type is `Int -> Int`.
* `f` is a function from a set of integers to a set of integers.
    Its type is `Set(Int) -> Set(Int)`.
* `r` is a record that has the fields `a` and `b`, where `a` is an integer
    and `b` is a string. Its type is `[a: Int, b: Str]`.
* `F` is a set of functions from a pair of integers to an integer.
   Its type is `Set(<<Int, Int>> -> Int)`.
* `Foo` is an operator of an integer and of a string that returns an integer.
   Its type is  `(Int, Str) => Int`.
* `Bar` is a higher-order operator that takes an operator that takes
    an integer and a string and returns an integer, and returns a Boolean.
   Its type is  `((Int, Str) => Int) => Bool`.
* `Baz` is a polymorphic operator that takes two arguments of the same type
    and returns a value of the type equal to the types of its arguments.
   Its type is `(a, a) => a`.
* `Proc` and `Faulty` are sets of the same type.

>    Their type is `Set(PID)`.

Reviewer: where does PID come from? What does it mean?

### 1.2. Type aliases

> We extend the type grammar of `T` with one more rule that introduces a type
> alias:

Reviewer: ```suggestion
We extend the type grammar of `T` with one more rule that introduces type
aliases:
```

Reviewer: This intro to the section doesn't make sense to me, because the type alias is specified in the grammar to begin with, not added as an extensions here (the way the current document is written). Perhaps just say "kthe grammer of `T` includes a rule" rather than "we extend"?

The type checker uses the type annotation to refine the type of an empty set
(or, of an empty sequence).

<a id="typeAlias"></a>

### 2.4. Introducing and using type aliases

Follow the four simple rules:

1. Type aliases can be only introduced at top-level declarations. That is, type aliases next to nested LET-IN
   definitions are ignored.

> 1. Assuming Rule 1, you can define a type alias with `@typeAlias` in the same position, where you could define `@type`.

Reviewer: ```suggestion
1. Assuming Rule 1, you can define a type alias with `@typeAlias` anywhere you can define a `@type`.
```
